arcjet/index.ts

```diff
@@ -2077,6 +2136,128 @@ export function protectSignup<const Characteristics extends string[] = []>(
   ];
 }
 
+/**
+ * Arcjet filter rule.
+ *
+ * @param options
+ *   Configuration (required).
+ * @returns
+ *   Filter rule.
+ */
+export function filter(options: FilterOptions): Primitive<{}> {
+  validateFilterOptions(options);
+
+  const mode = options.mode === "LIVE" ? "LIVE" : "DRY_RUN";
+  const allow = options.allow ?? [];
+  const deny = options.deny ?? [];
+
+  if (allow.length > 0 && deny.length > 0) {
+    throw new Error(
+      "`filter` options error: expressions must be passed in either `allow` or `deny` instead of both",
+    );
+  }
+  if (allow.length === 0 && deny.length === 0) {
+    throw new Error(
+      "`filter` options error: one or more expressions must be passed in `allow` or `deny`",
+    );
+  }
+
+  const state = mode === "LIVE" ? "RUN" : "DRY_RUN";
+  const type = "FILTER";
+  const version = 0;
+
+  const ruleIdPromise = hasher.hash(
+    hasher.string("type", type),
+    hasher.uint32("version", version),
+    hasher.string("mode", mode),
+    hasher.stringSliceOrdered(
+      "allow",
+      // @ts-expect-error: `hasher` must support readonly values.
+      allow,
+    ),
+    hasher.stringSliceOrdered(
+      "deny",
+      // @ts-expect-error: `hasher` must support readonly values.
+      deny,
+    ),
+  );
+
+  const rule: ArcjetFilterRule = {
+    allow,
+    deny,
+    mode,
+    priority: Priority.Filter,
+    async protect(
+      context: ArcjetContext<CachedResult>,
+      request: ArcjetRequestDetails,
+    ): Promise<ArcjetRuleResult> {
+      const ruleId = await ruleIdPromise;
+      const [cached, ttl] = await context.cache.get(
+        ruleId,
+        context.fingerprint,
+      );
+
+      if (cached) {
+        return new ArcjetRuleResult({
+          conclusion: cached.conclusion,
+          fingerprint: context.fingerprint,
+          reason: cached.reason,
+          ruleId,
+          state: "CACHED",
+          ttl,
+        });
+      }
+
+      const request_ = toAnalyzeRequest(request);
+      let ruleResult: ArcjetRuleResult;
+
+      try {
+        const result = await analyze.matchFilters(
+          context,
+          request_,
+          allow.length > 0 ? allow : deny,
+          allow.length > 0,
+        );
+
+        if (!result) {
+          throw new Error("WebAssembly is not supported in this runtime");
+        }
+
+        ruleResult = new ArcjetRuleResult({
+          conclusion: result.allowed ? "ALLOW" : "DENY",
+          fingerprint: context.fingerprint,
+          reason: new ArcjetFilterReason(result.matchedExpressions[0]),
+          ruleId,
+          state,
+          ttl: 60,
+        });
+      } catch (error) {
+        ruleResult = new ArcjetRuleResult({
+          conclusion: "ERROR",
+          fingerprint: context.fingerprint,
+          reason: new ArcjetErrorReason(String(error)),
```

Comment:
```suggestion
          reason: new ArcjetErrorReason(error),
```
ErrorReason accepts an error value from a catch

Reply:
done